File: Strategy Summit/Dynamic Programming/Q1. Climbing Stairs/Q1_ClimbingStairs.java
class Solution {
    public int climbStairs(int n) {
        if (n <= 2) return n; // base cases: 1 step -> 1 way, 2 steps -> 2 ways

        int first = 1;  // ways to reach step 1
        int second = 2; // ways to reach step 2
        int third = 0;  // ways to reach current step

        // Loop from step 3 to n
        for (int i = 3; i <= n; i++) {
            third = first + second; // ways to reach current step = sum of previous two steps
            first = second;         // shift first
            second = third;         // shift second
        }

        return third; // return ways to reach step n
    }
}

File: Strategy Summit/Dynamic Programming/Q1. Climbing Stairs/Q1_ClimbingStairs.py
class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n <= 2:
            return n  # base cases: 1 step -> 1 way, 2 steps -> 2 ways

        first = 1   # ways to reach step 1
        second = 2  # ways to reach step 2

        # Loop from step 3 to n
        for i in range(3, n + 1):
            third = first + second  # current ways = sum of previous two steps
            first = second          # shift first
            second = third          # shift second

        return second  # return ways to reach step n
